src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+                    }
+                }
+                else
+                {
+                    _workspace.ApplyChangeToWorkspace(w =>
+                    {
+                        var projectReference = _workspace.TryRemoveConvertedProjectReference(Id, fullPath, properties);
+
+                        // If this was converted to a project reference, we have now recorded the removal -- let's remove it here too
+                        if (projectReference != null)
+                        {
+                            w.OnProjectReferenceRemoved(Id, projectReference);
+                        }
+                        else
+                        {
+                            // TODO: find a cleaner way to fetch this
+                            var metadataReference = w.CurrentSolution.GetProject(Id).MetadataReferences.Cast<PortableExecutableReference>()
+                                                                                    .Single(m => m.FilePath == fullPath && m.Properties == properties);
+
+                            w.OnMetadataReferenceRemoved(Id, metadataReference);
+                        }
+                    });
+                }
+            }
+        }
+
+        #endregion
+
+        #region Project Reference Addition/Removal
+
+        public void AddProjectReference(ProjectReference projectReference)
+        {
+            if (projectReference == null)
+            {
+                throw new ArgumentNullException(nameof(projectReference));
+            }
+
+            lock (_gate)
+            {
+                if (ContainsProjectReference(projectReference))
+                {
+                    throw new ArgumentException("The project reference has already been added to the project.");
+                }
+
+                if (_activeBatchScopes > 0)
+                {
+                    if (!_projectReferencesRemovedInBatch.Remove(projectReference))
+                    {
+                        _projectReferencesAddedInBatch.Add(projectReference);
+                    }
+                }
+                else
+                {
+                    _workspace.ApplyChangeToWorkspace(w => w.OnProjectReferenceAdded(Id, projectReference));
+                }
+            }
+        }
+
+        public bool ContainsProjectReference(ProjectReference projectReference)
+        {
+            if (projectReference == null)
+            {
+                throw new ArgumentNullException(nameof(projectReference));
+            }
+
+            lock (_gate)
+            {
+                if (_projectReferencesRemovedInBatch.Contains(projectReference))
+                {
+                    return false;
+                }
+
+                if (_projectReferencesAddedInBatch.Contains(projectReference))
+                {
+                    return true;
+                }
+
+                return _workspace.CurrentSolution.GetProject(Id).AllProjectReferences.Contains(projectReference);
```

Comment:
>_workspace.CurrentSolution.GetProject(Id).AllProjectReferences.Contains(projectReference); [](start = 23, length = 90)

This can be done outside of the lock, if we capture the current solution inside the lock.